src/dataiterators/dataiterator.jl

```diff
@@ -1,30 +1,53 @@
-StatsBase.nobs(A::AbstractVector) = length(A)
-StatsBase.nobs(A::AbstractMatrix) = size(A, 2)
-StatsBase.nobs{T}(A::AbstractArray{T,3}) = size(A, 3)
-StatsBase.nobs{T}(A::AbstractArray{T,4}) = size(A, 4)
-
-getobs(A::AbstractVector, idx) = A[idx]
-getobs(A::AbstractMatrix, idx) = A[:, idx]
-getobs{T}(A::AbstractArray{T,3}, idx) = A[:, :, idx]
-getobs{T}(A::AbstractArray{T,4}, idx) = A[:, :, :, idx]
-
-getobs(A::Vector, range::Range) = slice(A, range)
-getobs{T}(A::SubArray{T,1}, range::Range) = slice(A, range)
-getobs(A::Matrix, range::Range) = sub(A, :, range)
-getobs{T}(A::SubArray{T,2}, range::Range) = sub(A, :, range)
-getobs{T}(A::Array{T,3}, range::Range) = sub(A, :, :, range)
-getobs{T}(A::SubArray{T,3}, range::Range) = sub(A, :, :, range)
-getobs{T}(A::Array{T,4}, range::Range) = sub(A, :, :, :, range)
-getobs{T}(A::SubArray{T,4}, range::Range) = sub(A, :, :, :, range)
+
+
+export
+    DataIterator,
+        SubsetIterator,
+        Batch,
+            SequentialBatch,
+        BatchIterator,
+            RandomBatches,
+    batches
+
+# # add support for arrays up to 4 dimensions
+# for N in 1:4
+#     @eval begin
+#         # size of last dimension
+#         LearnBase.nobs{T}(A::AbstractArray{T,$N}) = size(A, $N)
+#
+#         # # apply a view to the last dimension
+#         # LearnBase.getobs{T}(A::AbstractArray{T,$N}, idx) = view(A,  $(fill(:(:),N-1)...), idx)
+#     end
+# end
+
+@generated function LearnBase.nobs(A::AbstractArray)
+    T, N = A.parameters
+    :(size(A, $N))
+end
+
+# apply a view to the last dimension
+@generated function LearnBase.getobs(A::AbstractArray, idx)
+    T, N = A.parameters
+    @assert N > 0
+    if N == 1 && idx <: Integer
+        :(A[idx])
+    else
+        :(view(A,  $(fill(:(:),N-1)...), idx))
+    end
+end
```

Comment:
This does something different. I only used views for `Array` and `SubArray` and not for all `AbstractArray`. Is this change on purpose because of 0.5 improvements? What happens if it call it with a sparse array?